proposals/nnnn-bitwise-copyable.md

```diff
@@ -0,0 +1,270 @@
+If this is the developer's intent, they can explicitly conform the type.
+
+For `@frozen` types, however, `BitwiseCopyable` conformance will be inferred even when in a module built with library evolution.
+That's because the compiler can see that the type's fields are all `BitwiseCopyable` and knows that they will remain that way.
+For example, the compiler will infer a conformance of the following struct
+```swift
+@frozen
+public struct Coordinate3 {
+  var x: Int
+  var y: Int
+}
+```
+to `BitwiseCopyable`.
+
+## Effect on ABI stability
+
+The addition of the `BitwiseCopyable` constraint to either a type or a protocol in a library will not cause an ABI break for users.
+
+## Source compatibility
+
+This addition of a new protocol will not impact existing source code that does not use it.
+
+Removing the `BitwiseCopyable` marker from a type is source-breaking.
+As a result, future versions of Swift may conform additional existing types to `BitwiseCopyable`, but will not remove it from any type already conforming to `BitwiseCopyable`.
+
+## Effect on API resilience
+
+Adding a `BitwiseCopyable` constraint on a generic type will not cause an ABI break.
+As with any protocol, the additional constraint can cause a source break for users.
+
+## Future Directions
+
+### Automatic derivation of conditional conformances
+
+The wrapper type mentioned above
+```swift
+struct RegularBox<Value> {
+  var first: Value
+}
+```
+cannot conform to `BitwiseCopyable` unconditionally.
+It can, however, so long as `Value` is `BitwiseCopyable`.
+
+With this proposal, such a conditional conformance can be added manually:
+
+```swift
+extension Box : BitwiseCopyable where Value : BitwiseCopyable {}
+```
+
+In the future we may in some cases be able to derive it automatically.
+
+### MemoryLayout<T>.isBitwiseCopyable
+
+In certain circumstances, it would be useful to be able to dynamically determine whether a type conforms to `BitwiseCopyable`.
+In order to allow that, a new field could be added to `MemoryLayout`.
+
+### BitwiseMovable
+
+Most Swift types have the property that their representation can be relocated in memory with direct memory operations.
+This could be represented with a `BitwiseMovable` protocol that would be handled similarly to `BitwiseCopyable`.
+
+## Alternatives considered
+
+### Alternate Spellings
+
+**Trivial** is widely used within the compiler and Swift evolution discussions to refer to the property of bitwise copyability. `BitwiseCopyable`, on the other hand, is more self-documenting.
+
+## Acknowledgments
+
+This proposal has benefitted from discussions with John McCall, Joe Groff, Andrew Trick, Michael Gottesman, and Arnold Schwaigofer.
+
+## Appendix: Standard library conformers<a name="all-stdlib-conformers"/>
+
+The following protocols in the standard library will gain the `BitwiseCopyable` constraint:
+
+- `FixedWidthInteger`
```

Comment:
The language steering group discussed these proposed conformances and we feel that `FixedWidthInteger` shouldn't be in this list. While the concrete types in stdlib that conform to it are certainly bitwise copyable, there are no semantic requirements of the protocol itself that _require_ bitwise copyability.

The summary of `FixedWidthInteger` from the documentation is "An integer type that uses a fixed size for every instance"; while it's unlikely, an implementation _could_ do something a bit tortured like store a pointer to heap storage that contains the actual value. It was felt that the term "fixed width" here refers to the range of values it can represent, rather than the internal representation.

Would you agree/disagree with that?

Reply:
I would defer to @stephentyrone on this.  It doesn't seem to make a practical difference, since we're explicitly notating every existing fixed-width integer type.

Reply:
This remains unchanged for now, pending comment.

Reply:
@stephentyrone Can you comment quickly on this, please?

Reply:
What Tony said. Every FixedWidthInteger type in the stdlib is bitwise copyable, but it is technically possible for someone to implement a fixed-width integer type that is not.

Reply:
Removed.